x/erc20/keeper/ibc_callbacks_integration_test.go

```diff
@@ -0,0 +1,823 @@
+			s.Require().NoError(err)
+
+			// Check Balance
+			balanceToken :=
+				s.EvmosChain.App.(*app.Evmos).Erc20Keeper.BalanceOf(s.EvmosChain.GetContext(), contracts.ERC20MinterBurnerDecimalsContract.ABI, pair.GetERC20Contract(), common.BytesToAddress(senderAcc.Bytes()))
+			s.Require().Equal(amount, balanceToken.Int64())
+
+			// Attempt to send that will fail because balance is not enough
+			path := s.pathOsmosisEvmos
+			originEndpoint := path.EndpointB
+			originChain := s.EvmosChain
+			coin := pair.Erc20Address
+			transferMsg := transfertypes.NewMsgTransfer(originEndpoint.ChannelConfig.PortID, originEndpoint.ChannelID, sdk.NewCoin(coin, sdk.NewInt(amount*2)), sender, receiver, timeoutHeight, 0)
+
+			originChain.Coordinator.UpdateTimeForChain(originChain)
+
+			_, _, err = simapp.SignAndDeliver(
+				originChain.T,
+				originChain.TxConfig,
+				originChain.App.GetBaseApp(),
+				originChain.GetContext().BlockHeader(),
+				[]sdk.Msg{transferMsg},
+				originChain.ChainID,
+				[]uint64{originChain.SenderAccount.GetAccountNumber()},
+				[]uint64{originChain.SenderAccount.GetSequence()},
+				true, false, originChain.SenderPrivKey,
+			)
+
+			// Require a failing transfer
+			s.Require().Error(err)
+			// NextBlock calls app.Commit()
+			originChain.NextBlock()
+			originChain.Coordinator.IncrementTime()
+
+			// Check Balance didnt change
+			ibcOsmosBalance := s.IBCOsmosisChain.GetSimApp().BankKeeper.GetBalance(s.IBCOsmosisChain.GetContext(), receiverAcc, erc20Denomtrace.IBCDenom())
+			s.Require().Equal(int64(0), ibcOsmosBalance.Amount.Int64())
+			balanceToken =
+				s.EvmosChain.App.(*app.Evmos).Erc20Keeper.BalanceOf(s.EvmosChain.GetContext(), contracts.ERC20MinterBurnerDecimalsContract.ABI, pair.GetERC20Contract(), common.BytesToAddress(senderAcc.Bytes()))
+			s.Require().Equal(amount, balanceToken.Int64())
+		})
+	})
+	Describe("registered coin", func() {
+		BeforeEach(func() {
+			receiver = s.IBCOsmosisChain.SenderAccount.GetAddress().String()
+			sender = s.EvmosChain.SenderAccount.GetAddress().String()
+			receiverAcc = sdk.MustAccAddressFromBech32(receiver)
+			senderAcc = sdk.MustAccAddressFromBech32(sender)
+
+			erc20params := types.DefaultParams()
+			erc20params.EnableErc20 = false
+			s.EvmosChain.App.(*app.Evmos).Erc20Keeper.SetParams(s.EvmosChain.GetContext(), erc20params)
+
+			// Send from osmosis to Evmos
+			s.SendAndReceiveMessage(s.pathOsmosisEvmos, s.IBCOsmosisChain, "uosmo", amount, receiver, sender, 1, "")
+			s.EvmosChain.Coordinator.CommitBlock(s.EvmosChain)
+			erc20params.EnableErc20 = true
+			s.EvmosChain.App.(*app.Evmos).Erc20Keeper.SetParams(s.EvmosChain.GetContext(), erc20params)
+
+			// Register uosmo pair
+			var err error
+			pair, err = s.EvmosChain.App.(*app.Evmos).Erc20Keeper.RegisterCoin(s.EvmosChain.GetContext(), osmoMeta)
+			s.Require().NoError(err)
+		})
+		It("should transfer available balance", func() {
```

Comment:
should there be another case that says `"it should convert ERC20 to IBC voucher and transfer if the user doesn't have enough balance"`?